File: exams/frmResultAnalysis.vb
Public Class frmResultAnalysis

    Dim total_percentage As Double = 0
    Dim msg As String = ""
    Private Sub frmExamEntry_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        If Not connect() Then
            Me.Close()
        Else
            stream_mode = False
            lstExaminations.Items.Clear()
            CheckBox1.Enabled = True
            btnAddExam.Enabled = False
            txtContribution.Enabled = False
            grpMultiExaminations.Enabled = False
            fill_year()
            get_term()
            load_class(cboClass)
            cboTerm.SelectedItem = term
            cboYear.SelectedItem = Today.Year
        End If
    End Sub

    Private Sub loadClasses()
        cboClass.Items.Clear()
        qread("SELECT class FROM examined_classes WHERE (Examination='" & escape_string(cboExamName.SelectedItem) & "' AND Term='" & escape_string(cboTerm.SelectedItem) & "' AND Year='" & cboYear.SelectedItem & "')")
        While dbreader.Read
            cboClass.Items.Add(dbreader("class"))
        End While
    End Sub
    Private Sub fill_year()

        cboYear.Items.Clear()
        fill_years(cboYear)
        cboYear.SelectedItem = Today.Year
    End Sub
    Private Sub cboTerm_SelectedValueChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles cboTerm.SelectedValueChanged, cboYear.SelectedValueChanged
        If cboTerm.SelectedItem <> None And cboYear.SelectedItem.ToString <> None Then
            fill_exam()
        End If
    End Sub
    Private Sub fill_exam()
        If qread("SELECT ExamName FROM examinations WHERE Term='" & cboTerm.SelectedItem & "' AND Year='" & cboYear.SelectedItem & "'") Then
            cboExamName.Items.Clear()
            cboExamName.Items.Add(None)
            While dbreader.Read
                cboExamName.Items.Add(dbreader("ExamName"))
            End While
            cboExamName.SelectedItem = None
            dbreader.Close()
        Else
            failure("Could Not Load Examinations!")
        End If
    End Sub

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click
        search_teachers = False
        Me.Close()
    End Sub

    Private Sub btnClear_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnClear.Click
        clear()
        total_percentage = 0
    End Sub
    Private Sub clear()
        cboTerm.SelectedItem = None
        cboExamName.SelectedItem = None
        cboClass.SelectedItem = None
        'cboStream.SelectedItem = None
        lstExaminations.Items.Clear()
    End Sub
    Private Function isvalid()

        If total_percentage <> 100 And chkMode.Checked Then
            msg = "The Examination Contributions Don't Total to 100%!"
            Return False
        End If

        If Not IsNumeric(cboYear.SelectedItem) Then
            msg = "Invalid Selection For Year!"
            Return False
        ElseIf cboTerm.SelectedItem = None Then
            msg = "Invalid Selection for Term!"
            Return False
        ElseIf cboExamName.SelectedItem = None Then
            msg = "Invalid Selection For Exam Name!"
            Return False
        ElseIf cboClass.SelectedItem = Nothing Or cboClass.SelectedItem = None Then
            msg = "Invalid Selection For Class!"
            Return False
        ElseIf lstExaminations.Items.Count < 1 And chkMode.Checked Then
            msg = "Only 1 Examination? You didn't tell me so!"
            Return False
        ElseIf cboSortBy.SelectedItem = "" And cboSortBy.Visible Then
            msg = "Please Choose The Mode To Sort The Result Analysis!"
            Return False
        Else

            Return True
        End If

    End Function
    Private Sub load_entry_form_many()
        Dim i As Integer
        ReDim tables(lstExaminations.Items.Count - 1)
        ReDim contribution(lstExaminations.Items.Count - 1)
        ReDim total_mark(lstExaminations.Items.Count - 1)
        ReDim tms(lstExaminations.Items.Count - 1)
        ReDim yrs(lstExaminations.Items.Count - 1)
        For i = 0 To tables.Length - 1
            tables(i) = lstExaminations.Items.Item(i).Text
            total_mark(i) = get_total_mark(lstExaminations.Items.Item(i).Text, lstExaminations.Items.Item(i).SubItems(3).Text)
            contribution(i) = lstExaminations.Items.Item(i).SubItems(1).Text
            tms(i) = lstExaminations.Items(i).SubItems(3).Text
            yrs(i) = lstExaminations.Items(i).SubItems(2).Text
        Next
        class_form = get_name(cboClass.SelectedItem)
        ReDim exam_names(lstExaminations.Items.Count - 1)
        For i = 0 To lstExaminations.Items.Count - 1
            exam_names(i) = lstExaminations.Items.Item(i).Text
        Next
        table = "exam_results"
        For k As Integer = 0 To tables.Length - 1
            If total_mark(k) = 0 Then
                failure(" Program Cannot Perform The Requested Operation! Looks Like There's One Examination Which The Chosen Class Did Not Undertake!")
                Exit Sub
            End If
        Next
        Me.Close()
    End Sub
    Private Sub btnEnterMarks_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnEnterMarks.Click
        If isvalid() Then

            table = "exam_results"
            search_teachers = True
            show_split = CheckBox1.Checked
            tm = cboTerm.SelectedItem
            'stream = cboStream.SelectedItem
            yr = cboYear.SelectedItem
            If cboSortBy.SelectedItem = "Total Marks" Then
                sortby = "Total"
            ElseIf cboSortBy.SelectedItem = "Mean Marks" Then
                sortby = "MM"
            ElseIf cboSortBy.SelectedItem = "Mean Points" Then
                sortby = "MP"
            ElseIf cboSortBy.SelectedItem = "Total Points" Then
                sortby = "TP"
            End If

            If chkMode.Checked Then
                mode = True
                load_entry_form_many()
            Else
                chkMode.Checked = True
                mode = True
                lstExaminations.Items.Clear()
                Dim li As New ListViewItem
                li = lstExaminations.Items.Add(cboExamName.SelectedItem)
                li.SubItems.Add("100")
                li.SubItems.Add(cboYear.SelectedItem)
                li.SubItems.Add(cboTerm.SelectedItem)
                total_percentage += 100
                load_entry_form_many()
                ' load_entry_form()
            End If

        Else
            failure(msg)
        End If
    End Sub
    Private Sub load_entry_form()
        table = "exam_results"
        class_form = get_name(cboClass.SelectedItem)
        exam_name = cboExamName.SelectedItem
        Me.Close()
    End Sub

    Private Sub chkMode_CheckedChanged(ByVal sender As Object, ByVal e As System.EventArgs) Handles chkMode.CheckedChanged
        If chkMode.Checked Then
            btnAddExam.Enabled = True
            txtContribution.Enabled = True
            grpMultiExaminations.Enabled = True
            CheckBox1.Checked = False
            CheckBox1.Enabled = False
        Else
            lstExaminations.Items.Clear()
            CheckBox1.Enabled = True
            btnAddExam.Enabled = False
            txtContribution.Enabled = False
            grpMultiExaminations.Enabled = False
        End If

    End Sub
    Private Function exists(ByVal str As String)
        Dim i As Integer
        Dim selectedTerm As String = String.Empty

        For i = 0 To lstExaminations.Items.Count - 1
            If lstExaminations.Items.Item(i).Text = str And lstExaminations.Items.Item(i).SubItems(2).Text = cboYear.SelectedItem And lstExaminations.Items.Item(i).SubItems(3).Text = cboTerm.SelectedItem Then
                MsgBox("The Exam Already Exist In The List")
                Return True
            End If
        Next
        Return False
    End Function
    Private Sub add_exam()
        If cboExamName.SelectedItem <> None And IsNumeric(txtContribution.Text) And txtContribution.Text <> "" Then
            If Not exists(cboExamName.SelectedItem) Then
                Dim li As New ListViewItem
                li = lstExaminations.Items.Add(cboExamName.SelectedItem)
                li.SubItems.Add(txtContribution.Text)
                li.SubItems.Add(cboYear.SelectedItem)
                li.SubItems.Add(cboTerm.SelectedItem)
                total_percentage += txtContribution.Text
            End If
        End If
    End Sub
    Private Sub btnAddExam_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnAddExam.Click
        add_exam()
    End Sub

    Private Sub cboExamName_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cboExamName.SelectedIndexChanged
        txtContribution.Clear()
        txtContribution.Focus()
        loadClasses()
    End Sub

    Private Sub txtContribution_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtContribution.KeyPress
        If Asc(e.KeyChar()) = 13 Then
            add_exam()
        End If
    End Sub

    Private Sub cboClass_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cboClass.SelectedIndexChanged
        get_class_subjects(cboClass.SelectedItem)
    End Sub
End Class

File: exams/frmResultAnalysis.cs
using System;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace exams
{
    public partial class frmResultAnalysis
    {
        public frmResultAnalysis()
        {
            InitializeComponent();
            _btnAddExam.Name = "btnAddExam";
            _chkMode.Name = "chkMode";
            _btnEnterMarks.Name = "btnEnterMarks";
            _btnClear.Name = "btnClear";
            _txtContribution.Name = "txtContribution";
            _btnCancel.Name = "btnCancel";
            _cboClass.Name = "cboClass";
            _cboExamName.Name = "cboExamName";
            _cboYear.Name = "cboYear";
            _cboTerm.Name = "cboTerm";
        }

        private double total_percentage = 0d;
        private string msg = string.Empty;

        private void frmExamEntry_Load(object sender, EventArgs e)
        {
            if (!publicSubsNFunctions.connect())
            {
                Close();
            }
            else
            {
                publicSubsNFunctions.stream_mode = false;
                lstExaminations.Items.Clear();
                CheckBox1.Enabled = true;
                btnAddExam.Enabled = false;
                txtContribution.Enabled = false;
                grpMultiExaminations.Enabled = false;
                fill_year();
                publicSubsNFunctions.get_term();
                var argcbo = cboClass;
                publicSubsNFunctions.load_class(ref argcbo);
                cboClass = argcbo;
                cboTerm.SelectedItem = publicSubsNFunctions.term;
                cboYear.SelectedItem = DateAndTime.Today.Year;
            }
        }

        private void loadClasses()
        {
            cboClass.Items.Clear();
            string argq = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject("SELECT class FROM examined_classes WHERE (Examination='" + publicSubsNFunctions.escape_string(Conversions.ToString(cboExamName.SelectedItem)) + "' AND Term='" + publicSubsNFunctions.escape_string(Conversions.ToString(cboTerm.SelectedItem)) + "' AND Year='", cboYear.SelectedItem), "')"));
            publicSubsNFunctions.qread(ref argq);
            while (publicSubsNFunctions.dbreader.Read())
                cboClass.Items.Add(publicSubsNFunctions.dbreader["class"]);
        }

        private void fill_year()
        {
            cboYear.Items.Clear();
            var argcbo = cboYear;
            publicSubsNFunctions.fill_years(ref argcbo);
            cboYear = argcbo;
            cboYear.SelectedItem = DateAndTime.Today.Year;
        }

        private void cboTerm_SelectedValueChanged(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false), (cboYear.SelectedItem.ToString() ?? "") != (publicSubsNFunctions.None ?? ""))))
            {
                fill_exam();
            }
        }

        private void fill_exam()
        {
            string argq = Conversions.ToString(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject(Operators.ConcatenateObject("SELECT ExamName FROM examinations WHERE Term='", cboTerm.SelectedItem), "' AND Year='"), cboYear.SelectedItem), "'"));
            if (publicSubsNFunctions.qread(ref argq))
            {
                cboExamName.Items.Clear();
                cboExamName.Items.Add(publicSubsNFunctions.None);
                while (publicSubsNFunctions.dbreader.Read())
                    cboExamName.Items.Add(publicSubsNFunctions.dbreader["ExamName"]);
                cboExamName.SelectedItem = publicSubsNFunctions.None;
                publicSubsNFunctions.dbreader.Close();
            }
            else
            {
                publicSubsNFunctions.failure("Could Not Load Examinations!");
            }
        }

        private void btnCancel_Click(object sender, EventArgs e)
        {
            publicSubsNFunctions.search_teachers = false;
            Close();
        }

        private void btnClear_Click(object sender, EventArgs e)
        {
            clear();
            total_percentage = 0d;
        }

        private void clear()
        {
            cboTerm.SelectedItem = publicSubsNFunctions.None;
            cboExamName.SelectedItem = publicSubsNFunctions.None;
            cboClass.SelectedItem = publicSubsNFunctions.None;
            // cboStream.SelectedItem = None
            lstExaminations.Items.Clear();
        }

        private object isvalid()
        {
            if (total_percentage != 100d & chkMode.Checked)
            {
                msg = "The Examination Contributions Don't Total to 100%!";
                return false;
            }

            if (!Information.IsNumeric(cboYear.SelectedItem))
            {
                msg = "Invalid Selection For Year!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboTerm.SelectedItem, publicSubsNFunctions.None, false)))
            {
                msg = "Invalid Selection for Term!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboExamName.SelectedItem, publicSubsNFunctions.None, false)))
            {
                msg = "Invalid Selection For Exam Name!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.OrObject(Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, null, false), Operators.ConditionalCompareObjectEqual(cboClass.SelectedItem, publicSubsNFunctions.None, false))))
            {
                msg = "Invalid Selection For Class!";
                return false;
            }
            else if (lstExaminations.Items.Count < 1 & chkMode.Checked)
            {
                msg = "Only 1 Examination? You didn't tell me so!";
                return false;
            }
            else if (Conversions.ToBoolean(Operators.AndObject(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, string.Empty, false), cboSortBy.Visible)))
            {
                msg = "Please Choose The Mode To Sort The Result Analysis!";
                return false;
            }
            else
            {
                return true;
            }
        }

        private void load_entry_form_many()
        {
            int i;
            publicSubsNFunctions.tables = new string[lstExaminations.Items.Count];
            publicSubsNFunctions.contribution = new int[lstExaminations.Items.Count];
            publicSubsNFunctions.total_mark = new int[lstExaminations.Items.Count];
            publicSubsNFunctions.tms = new object[lstExaminations.Items.Count];
            publicSubsNFunctions.yrs = new object[lstExaminations.Items.Count];
            var loopTo = publicSubsNFunctions.tables.Length - 1;
            for (i = 0; i <= loopTo; i++)
            {
                publicSubsNFunctions.tables[i] = lstExaminations.Items[i].Text;
                publicSubsNFunctions.total_mark[i] = Conversions.ToInteger(publicSubsNFunctions.get_total_mark(lstExaminations.Items[i].Text, lstExaminations.Items[i].SubItems[3].Text));
                publicSubsNFunctions.contribution[i] = Conversions.ToInteger(lstExaminations.Items[i].SubItems[1].Text);
                publicSubsNFunctions.tms[i] = lstExaminations.Items[i].SubItems[3].Text;
                publicSubsNFunctions.yrs[i] = lstExaminations.Items[i].SubItems[2].Text;
            }

            publicSubsNFunctions.class_form = Conversions.ToString(publicSubsNFunctions.get_name(Conversions.ToString(cboClass.SelectedItem)));
            publicSubsNFunctions.exam_names = new object[lstExaminations.Items.Count];
            var loopTo1 = lstExaminations.Items.Count - 1;
            for (i = 0; i <= loopTo1; i++)
                publicSubsNFunctions.exam_names[i] = lstExaminations.Items[i].Text;
            publicSubsNFunctions.table = "exam_results";
            for (int k = 0, loopTo2 = publicSubsNFunctions.tables.Length - 1; k <= loopTo2; k++)
            {
                if (publicSubsNFunctions.total_mark[k] == 0)
                {
                    publicSubsNFunctions.failure(" Program Cannot Perform The Requested Operation! Looks Like There's One Examination Which The Chosen Class Did Not Undertake!");
                    return;
                }
            }

            Close();
        }

        private void btnEnterMarks_Click(object sender, EventArgs e)
        {
            if (Conversions.ToBoolean(isvalid()))
            {
                publicSubsNFunctions.table = "exam_results";
                publicSubsNFunctions.search_teachers = true;
                publicSubsNFunctions.show_split = CheckBox1.Checked;
                publicSubsNFunctions.tm = Conversions.ToString(cboTerm.SelectedItem);
                // stream = cboStream.SelectedItem
                publicSubsNFunctions.yr = Conversions.ToInteger(cboYear.SelectedItem);
                if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Total Marks", false)))
                {
                    publicSubsNFunctions.sortby = "Total";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Mean Marks", false)))
                {
                    publicSubsNFunctions.sortby = "MM";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Mean Points", false)))
                {
                    publicSubsNFunctions.sortby = "MP";
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(cboSortBy.SelectedItem, "Total Points", false)))
                {
                    publicSubsNFunctions.sortby = "TP";
                }

                if (chkMode.Checked)
                {
                    publicSubsNFunctions.mode = true;
                    load_entry_form_many();
                }
                else
                {
                    chkMode.Checked = true;
                    publicSubsNFunctions.mode = true;
                    lstExaminations.Items.Clear();
                    var li = new ListViewItem();
                    li = lstExaminations.Items.Add(cboExamName.SelectedItem);
                    li.SubItems.Add("100");
                    li.SubItems.Add(cboYear.SelectedItem);
                    li.SubItems.Add(cboTerm.SelectedItem);
                    total_percentage += 100d;
                    load_entry_form_many();
                    // load_entry_form()
                }
            }
            else
            {
                publicSubsNFunctions.failure(msg);
            }
        }

        private void load_entry_form()
        {
            publicSubsNFunctions.table = "exam_results";
            publicSubsNFunctions.class_form = Conversions.ToString(publicSubsNFunctions.get_name(Conversions.ToString(cboClass.SelectedItem)));
            publicSubsNFunctions.exam_name = Conversions.ToString(cboExamName.SelectedItem);
            Close();
        }

        private void chkMode_CheckedChanged(object sender, EventArgs e)
        {
            if (chkMode.Checked)
            {
                btnAddExam.Enabled = true;
                txtContribution.Enabled = true;
                grpMultiExaminations.Enabled = true;
                CheckBox1.Checked = false;
                CheckBox1.Enabled = false;
            }
            else
            {
                lstExaminations.Items.Clear();
                CheckBox1.Enabled = true;
                btnAddExam.Enabled = false;
                txtContribution.Enabled = false;
                grpMultiExaminations.Enabled = false;
            }
        }

        private object exists(string str)
        {
            int i;
            string selectedTerm = string.Empty;
            var loopTo = lstExaminations.Items.Count - 1;
            for (i = 0; i <= loopTo; i++)
            {
                if (Conversions.ToBoolean(Operators.AndObject(Operators.AndObject((lstExaminations.Items[i].Text ?? "") == (str ?? ""), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[2].Text, cboYear.SelectedItem, false)), Operators.ConditionalCompareObjectEqual(lstExaminations.Items[i].SubItems[3].Text, cboTerm.SelectedItem, false))))
                {
                    Interaction.MsgBox("The Exam Already Exist In The List");
                    return true;
                }
            }

            return false;
        }

        private void add_exam()
        {
            if (Conversions.ToBoolean(Operators.AndObject(Operators.AndObject(Operators.ConditionalCompareObjectNotEqual(cboExamName.SelectedItem, publicSubsNFunctions.None, false), Information.IsNumeric(txtContribution.Text)), (txtContribution.Text ?? "") != (string.Empty ?? ""))))
            {
                if (Conversions.ToBoolean(!exists(Conversions.ToString(cboExamName.SelectedItem))))
                {
                    var li = new ListViewItem();
                    li = lstExaminations.Items.Add(cboExamName.SelectedItem);
                    li.SubItems.Add(txtContribution.Text);
                    li.SubItems.Add(cboYear.SelectedItem);
                    li.SubItems.Add(cboTerm.SelectedItem);
                    total_percentage += Conversions.ToDouble(txtContribution.Text);
                }
            }
        }

        private void btnAddExam_Click(object sender, EventArgs e)
        {
            add_exam();
        }

        private void cboExamName_SelectedIndexChanged(object sender, EventArgs e)
        {
            txtContribution.Clear();
            txtContribution.Focus();
            loadClasses();
        }

        private void txtContribution_KeyPress(object sender, KeyPressEventArgs e)
        {
            if (Strings.Asc(e.KeyChar) == 13)
            {
                add_exam();
            }
        }

        private void cboClass_SelectedIndexChanged(object sender, EventArgs e)
        {
            publicSubsNFunctions.get_class_subjects(Conversions.ToString(cboClass.SelectedItem));
        }
    }
}